src/treelearner/serial_tree_learner.cpp

```diff
@@ -521,10 +537,15 @@ void SerialTreeLearner::ConstructHistograms(const std::vector<int8_t>& is_featur
   #endif
 }
 
-void SerialTreeLearner::FindBestSplitsFromHistograms(const std::vector<int8_t>& is_feature_used, bool use_subtract) {
+void SerialTreeLearner::FindBestSplitsFromHistograms(
+    const std::vector<int8_t> &is_feature_used, bool use_subtract,
+    const Tree *tree) {
   #ifdef TIMETAG
   auto start_time = std::chrono::steady_clock::now();
   #endif
+    LearnerState learner_state(config_, data_partition_.get(), train_data_,
+                               constraints_per_leaf_, tree, current_constraints,
+                               cegb_);
```

Comment:
can this only be called one-time? not everytime in `FindBestSplitsFromHistograms`.

Reply:
BTW, could we avoid the extra calls in serial tree learner, when mc is not sed?

Reply:
> can this only be called one-time? not everytime in `FindBestSplitsFromHistograms`.

@guolinke I don't really understand what you mean here. If I am not mistaken the Learner state class is created once an only once every time we look for splits, so in a tree this will be called as many times as there are splits. This also a very fast call, because it is just setting references. So I don't understand why we would want to call it less than that. If we do, I as least need to call it once per tree, but then we would have "duplicated pointers" (pointers both in SerialTreeLearner and in LearnerState pointing to the same objects like config, data_partition_ or tree...) all the time, which I am not sure is a good idea. In light of this can you please clarify how you would like the code to be modified? Thanks  
@aldanor can you help me with that also if you have a good idea of how to improve the code?

Reply:
> BTW, could we avoid the extra calls in serial tree learner, when mc is not sed?

Given how we transformed the code now, I don't think we can avoid this call, otherwise we would have to duplicate the functions that use learner_state. Do you have a specific idea in mind of how you would like this to work?